keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,519 @@
+proxy the traffic to, otherwise it return an “Unknown network” error.
+
+The KAS starts with a proxy configuration like the below example.
+The example specifies 4 networks. "direct" specifies the KAS talking directly
+on the local network (no proxy). "master" specifies the KAS talks to a proxy
+listening at 1.2.3.4:5678. "cluster" specifies the KAS talk to a proxy
+listening at 1.2.3.5:5679. While these are represented as resources
+they are not intended to be loaded dynamically.
+The KAS loads them as configuration at start time.
+
+```yaml
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: direct
+  connection:
+    type: direct
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Master
+  connection:
+    type: grpc
+    url: grpc://1.2.3.4:5678
+    caBundle: file1.pem
+    clientKeyFile: proxy-client1.key
+    clientCertFile: proxy-client1.crt
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Cluster
+  connection:
+    type: grpc
+    url: grpc://1.2.3.5:5679
+    caBundle: file2.pem
+    clientKeyFile: proxy-client2.key
+    clientCertFile: proxy-client2.crt
+```
+
+NetworkContext could be extended to contain more contextual information.
+This would allow smarter routing based on the k8s object KAS is processing
+or which user/tenant tries to initiate the request, etc.
+
+### Proxy gRPC definition
+
+In order to serve a proxy request, one gRPC bidirectional stream on proxy
+server is created to serve it. It's a 1:1 mapping from TCP connection to a
+gRPC stream, so the state of TCP connection is exactly the same as the gRPC
+stream state.
+
+```grpc
+syntax = "proto3";
+
+service ProxyService {
+  // Proxy a TCP connection to a remote address defined by ConnectParam.
+  // The ConnectParam is defined in metadata under key "x-kube-net-proxy".
+  // metadata["x-kube-net-proxy"] = base64.Encode(proto.Marshal(connectOptions))
+  rpc Proxy(stream Payload) returns (stream Payload) {}
+}
+
+// ConnectOptions defines the remote TCP endpoint to connect
+message ConnectOptions {
+  string remote_addr = 1; // remote address to connect to. e.g. 8.8.8.8:53
+}
+
+// Payload defines a TCP payload.
+message Payload {
+  bytes data = 1;
+}
+```
+
+### Connectivity Proxy
+
+The connectivity proxy(s) can run in the same container as the KAS.
+It should run on the same machine and must run in the same flat network as the KAS.
+It listens on a port for gRPC connections from the KAS.
+This port would be for forwarding traffic to the appropriate cluster.
+It should have an administrative port speaking https.
+The administrative port serves the liveness probe and metrics.
+The liveness probe prevents a partially broken cluster
+where the KAS cannot connect to the cluster. This port also serves
+pprof debug commands and monitoring data for the proxy.
+
```

Comment:
early on the discussion was, allow the api-server to serve requests even if the proxy service. How will this work if an api-server instance is `healthy` yet the `connectivity proxy` isn't? We will get weird behavior where some api server calls succeed and some fail depending on which instance picked up the request and the type of that particular request. I believe we either find a way to load balance api-server --> connectivity proxy or we tie the health of api-server to the proxy by design (this will introduce a catch-22 situation where all nodes are unhealthy taking proxies down, and accordingly api-server).

Reply:
Good point. A few thoughts. 1) Proxy service should have a liveness check which means the kubelet shoudl be getting it to recover. 2) The proxy service is significantly simpler in scope, responsibility and code than the API Server, so it should be easier to make it more resilient. 3) In a HA environment we should absolutely consider having the API Servers readiness probe dependent on the proxy and thus control which API servers get traffic. 4) The design deliberately suggests configuration options rather than mandating them. A provider is welcome to  load-balance api-server to connectivity proxy.

Reply:
I agree it makes sense to operate the proxy as an HA LB service. I don't think the default setup needs to do that.

It's important to note that if this proxy is down, the cluster is only partially broken. Only master-> node calls stop working. Which is not a large fraction of requests. So, I don't actually agree that apiserver's readiness probe should depend on the proxy's readiness: that would make a minor problem into a cluster-breaking problem.

Reply:
> So, I don't actually agree that apiserver's readiness probe should depend on the proxy's readiness: that would make a minor problem into a cluster-breaking problem.

+1, that sounds like a recipe for cascading failure

Reply:
Thanks. I Agree, let us decouple the health of api-server from the proxy. 

The HA scenario is quite interesting. i would love to see the details around the proposed nodes->proxy connectivity as requested in https://github.com/kubernetes/enhancements/pull/872#discussion_r271859152